CMakeLists.txt

```diff
@@ -1,170 +1,89 @@
-  add_compile_options(-Wall -Werror -Wextra)
-
-  # clang-cl treats '-Wall' like '-Weverything' currently; so there are a few
-  # warnings we need to manually disable.
-
-  # Disable unnecessary compatibility and 'newline-eof'. This is a modern C++
-  # library -- so these serve no purpose
-  add_compile_options(-Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-newline-eof)
-
-  # Disable warnings on padding. This is a great warning, but not good when
-  # coupled with -Werror
-  add_compile_options(-Wno-padded)
-
-  # Several utilities require static objects that have exit-time destructors,
-  # including 'error_category' types from the standard.
-  add_compile_options(-Wno-exit-time-destructors)
+# ---- Warning guard ----
 
-  # This is erroring on valid uses of the standard library, like std::lock_guard.
-  add_compile_options(-Wno-ctad-maybe-unsupported)
+# Protect dependents from this project's warnings if the guard isn't disabled
+set(Result_warning_guard SYSTEM)
+if(is_top_project OR Result_INCLUDE_WITHOUT_SYSTEM)
+  set(Result_warning_guard "")
+endif()
 
-  # This is a good warning *in general*, but not when using MSVC's standard
-  # library, which requires defining "_" prefixed macros just to have it operate
-  # like the C++ standard dictates.
-  add_compile_options(-Wno-reserved-id-macro)
+# ---- Declare library ----
 
-  # Don't warn on ignored attributes; this breaks builds using dllexport/dllimport on
-  # classes that may also define inline functions
-  add_compile_options(-Wno-ignored-attributes)
+add_library(Result_Result INTERFACE)
+add_library(Result::Result ALIAS Result_Result)
 
-elseif ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang" OR
-        "${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
-  add_compile_options(-Wall -Werror -Wextra -pedantic -pedantic-errors)
+set_property(TARGET Result_Result PROPERTY EXPORT_NAME Result)
 
-  # Xcode 11.2 seems to enable -Wnewline-eof with '-Wextra'. Disable this,
-  # since it's a legacy requirement not needed in C++11 onward.
-  add_compile_options(-Wno-newline-eof)
-
-  # Don't warn on ignored attributes; this breaks builds using dllexport/dllimport on
-  # classes that may also define inline functions
-  add_compile_options(-Wno-ignored-attributes)
-elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
-  add_compile_options(/W4 /WX)
-endif ()
-
-include(AddSelfContainmentTest)
-
-if (RESULT_COMPILE_UNIT_TESTS)
-  add_self_containment_test(${PROJECT_NAME}.SelfContainmentTest
-    TARGET ${PROJECT_NAME}
-    HEADERS ${header_files}
-  )
-
-  add_subdirectory("test")
-endif ()
+target_include_directories(
+    Result_Result ${Result_warning_guard}
+    INTERFACE "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>"
+)
 
-##############################################################################
-# Installation
-##############################################################################
+target_compile_features(Result_Result INTERFACE cxx_std_11)
```

Comment:
Compile features exist [in 3.4](https://cmake.org/cmake/help/v3.4/command/target_compile_features.html). Please see my comment to the first conversation. `(CMAKE_)?<LANG>_STANDARD` is inadequate due to the decaying behavior. It's only use in modern CMake is to constrain the standard flag to a specific version from the CLI/a superbuild CML.

Reply:
I know that compile-features exist, but `std_cxx_11` doesn't in 3.4 -- meaning that this would require enumerating all of the features that are used in the library (which I am not a huge fan of doing, though this is an option).

> due to the decaying behavior

I'm not sure I understand what you mean by decaying behavior, or how this constrains the build. I may have to look into this

Reply:
`cxx_std_xx` are just meta features that explicitly state the standard. You can use [actual features](https://cmake.org/cmake/help/v3.4/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html) to approximate the C++11 requirement. There are [projects](https://github.com/nlohmann/json/blob/develop/CMakeLists.txt#L77-L78) already do this.

See [CXX_STANDARD](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html) for the decaying behavior.